feat(discover): Add a compare_aggregate function

wmak opened the pull request:
- This is primarily so we can do compare(p75>treshold), but I'm trying
  to set this up so we can use it in the future as well
- Sigh updated, to just `compare_numeric_aggregate`, will need more work to change this so its compatible regardless of the aggregate's type
- We need 2 compare functions here
  - When the thing being compared is an aggregate, we need this function
    to be treated as an aggregate so conditions etc. still work
  - Similarily for columns

wmak committed:
feat(discover): Add a compare_aggregate function

- This is primarily so we can do compare(p75>treshold), but I'm trying
  to set this up so we can use it in the future as well
- Overlaps a bit with #22091, I'm thinking after that goes in, to update
  it to be `compare_column`
- We'll need 2 compare functions here
  - When the thing being compared is an aggregate, we need this function
    to be treated as an aggregate so conditions etc. still work
  - Similarily for columns

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -2021,6 +2047,26 @@ def result_type_fn(function_arguments, parameter_values):
+            ],
+            aggregate=[
+                # snuba json syntax isn't compatible with this query here
+                # can't be a column, since we want to use this with aggregates
```
```suggestion
                # this function can't be a column, since we want to use this with aggregates
```

markstory commented on src/sentry/api/event_search.py:
```diff
@@ -2021,6 +2047,26 @@ def result_type_fn(function_arguments, parameter_values):
+            aggregate=[
+                # snuba json syntax isn't compatible with this query here
+                # can't be a column, since we want to use this with aggregates
+                "if({condition}({aggregate_alias},{value}),{pass},{fail})",
```
Do we need to ensure that the aggregate_alias is safe? Right now this looks like we could be open to naughty queries.

wmak replied:
going to use the same regex as #22129 for this, also updated the other functions that string substitute too

wmak committed:
feat: Add a function arg that validates aliases

- This uses the same regex pattern as #22129

wmak committed:
Update src/sentry/api/event_search.py

Co-authored-by: Mark Story <mark@sentry.io>

wmak committed:
Merge branch 'master' into feat/add-compare-aggregate

wmak committed:
ref: Changing this to specifically be for numeric aggregates

- will require far more work that I can fit in right now so that this
  works across different aggregate types

wmak committed:
ref: Change to a boolean result instead of string

- This means we sort a lot more rationally

k-fish commented on tests/snuba/api/endpoints/test_organization_events_v2.py:
```diff
@@ -2733,6 +2733,30 @@ def test_measurements_aggregation_conditions(self):
+
+        assert response.status_code == 200, response.content
+        assert len(response.data["data"]) == 1
+        assert response.data["data"][0]["compare_numeric_aggregate_p75_measurements_fcp_>_0"] == 1
```
nice, ints 👏

Zylphrex commented on src/sentry/api/event_search.py:
```diff
@@ -1418,6 +1418,7 @@ def get_json_meta_type(field_alias, snuba_type, function=None):
 
 
 FUNCTION_PATTERN = re.compile(r"^(?P<function>[^\(]+)\((?P<columns>[^\)]*)\)$")
+ALIAS_PATTERN = re.compile(r"(\w+)?(?!\d+)\w+$")
```
I don't think this impacts this PR right now, but the aliases can now contain comparison operators which would not match this pattern.

Zylphrex reviewed (approved):
Suggestion: use words instead of symbols for the comparison operators

wmak committed:
ref: Use strings for the conditions

- This is so that we don't end up with special symbols in our aliases

wmak commented:
> Suggestion: use words instead of symbols for the comparison operators
Good call, I've updated to strings

wmak committed:
Merge branch 'master' into feat/add-compare-aggregate

Zylphrex reviewed (approved):
🥳
